subprojects/gradle/runner/device-provider/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClient.kt

```diff
@@ -69,44 +75,46 @@ internal class KubernetesReservationClient(
         )
     }
 
-    private fun CoroutineScope.createDeployment(
+    private suspend fun createDeployment(
         reservation: ReservationData,
         deploymentsChannel: Channel<String>,
         podsChannel: Channel<Pod>,
         serialsChannel: Channel<DeviceCoordinate>
     ) {
-        launch {
-            val deployment = reservationDeploymentFactory.createDeployment(
-                namespace = kubernetesApi.namespace,
-                reservation = reservation
-            )
-            val deploymentName = deployment.metadata.name
-            deploymentsChannel.send(deploymentName)
-            kubernetesApi.createDeployment(deployment)
-
-            listenPodsFromDeployment(
-                deploymentName = deploymentName,
-                podsChannel = podsChannel,
-                serialsChannel = serialsChannel
-            )
-        }
+        val deployment = reservationDeploymentFactory.createDeployment(
+            namespace = kubernetesApi.namespace,
+            reservation = reservation
+        )
+        val deploymentName = deployment.metadata.name
+        deploymentsChannel.send(deploymentName)
+        kubernetesApi.createDeployment(deployment)
+
+        listenPodsFromDeployment(
+            deploymentName = deploymentName,
+            podsChannel = podsChannel,
+            serialsChannel = serialsChannel
+        )
     }
 
     private fun CoroutineScope.initializeDevices(
         podsChannel: ReceiveChannel<Pod>,
         serialsChannel: Channel<DeviceCoordinate>
     ) {
-        launch {
-            @Suppress("DEPRECATION") // todo use Flow
-            val uniqueRunningPods: ReceiveChannel<Pod> = podsChannel
+        launch(CoroutineName("waiting-pods")) {
+            @Suppress("DEPRECATION")
+            podsChannel
                 .filter { it.status.phase == POD_STATUS_RUNNING }
                 .distinctBy { it.metadata.name }
-
-            for (pod in uniqueRunningPods) {
-                launch {
-                    pod.bootDevice(serialsChannel)
+                .consumeEach { pod ->
+                    if (!serialsChannel.isClosedForSend) {
```

Comment:
if `serialsChannel` is closed, for example from another part of a runner then we stop consuming pods